tree.go

```diff
@@ -22,33 +22,56 @@ type Tree interface {
 	Root() ([]byte, error)
 }
 
-var _ Tree = &DefaultTree{}
+// treeFns is a global map used for keeping track of registered tree constructors for JSON serialization
+// The keys of this map should be kebab cased. E.g. "default-tree"
+var treeFns = sync.Map{}
+
+// Must be called in the init function
+func RegisterTree(treeName string, treeConstructor TreeConstructorFn) error {
+	if _, ok := treeFns.Load(treeName); ok {
+		return fmt.Errorf("%s already registered", treeName)
+	}
+
+	treeFns.Store(treeName, treeConstructor)
 
-type DefaultTree struct {
-	*merkletree.Tree
-	leaves [][]byte
-	root   []byte
+	return nil
 }
 
-func NewDefaultTree(_ Axis, _ uint) Tree {
-	return &DefaultTree{
-		Tree:   merkletree.New(sha256.New()),
-		leaves: make([][]byte, 0, 128),
+// Get tree constructor function by tree name from the global map registry
+func TreeFn(treeName string) (TreeConstructorFn, error) {
+	var treeFn TreeConstructorFn
+	v, ok := treeFns.Load(treeName)
+	if !ok {
+		return nil, fmt.Errorf("%s not registered yet", treeName)
+	}
+	treeFn, ok = v.(TreeConstructorFn)
+	if !ok {
+		return nil, fmt.Errorf("key %s has invalid interface", treeName)
 	}
-}
 
-func (d *DefaultTree) Push(data []byte) error {
-	// ignore the idx, as this implementation doesn't need that info
-	d.leaves = append(d.leaves, data)
-	return nil
+	return treeFn, nil
 }
 
-func (d *DefaultTree) Root() ([]byte, error) {
-	if d.root == nil {
-		for _, l := range d.leaves {
-			d.Tree.Push(l)
+// Get the tree name by the tree constructor function from the global map registry
+func getTreeNameFromConstructorFn(treeConstructor TreeConstructorFn) string {
```

Comment:
Let's leave here a little todo, saying it is temporary and ideally pointing to another PR/issue that makes the breaking changes for constructors and removes it. It would be nice to prepare that PR already.

Reply:
gotcha, i will draft that PR and link it in the TODO